keps/sig-ui/0035-add-plugin-support-to-dashboard.md

```diff
@@ -0,0 +1,154 @@
+It should be possible to collect this information before implementation begins in order to avoid requiring implementors to split their attention between writing release notes and implementing the feature itself.
+KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.
+
+A good summary is probably at least a paragraph in length. -->
+
+We want to introduce a plugin mechanism in dashboard which shall enable easy functionality addition to the existing dashboard. 
+
+## Motivation
+
+<!-- This section is for explicitly listing the motivation, goals and non-goals of this KEP.
+Describe why the change is important and the benefits to users.
+The motivation section can optionally provide links to [experience reports][] to demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports -->
+
+The idea was proposed back in 2017 [here](https://github.com/kubernetes/dashboard/issues/1832). This mechanism will also be helpful in better utilisation of [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) through the dashboard.
+
+A custom UI requirement came up for [ingress-nginx](https://github.com/kubernetes/ingress-nginx/issues/109) which later required [enhancement](https://github.com/kubernetes/ingress-nginx/issues/2480) and dashboard should be able to support it without requiring constant changes to `kubernetes/dashboard`.
+
+In general, users should be able to add functionality without modifying the dashboard source code.
+
+### Goals
+
+<!-- List the specific goals of the KEP.
+How will we know that this has succeeded? -->
+
+- A user should be able to install a custom plugin to the dashboard extending its functionality in some way.
+- We need to provide an interface to the plugin so that it can interact with core dashboard services.
+
+<!-- ### Non-Goals -->
+
+<!-- What is out of scope for his KEP?
+Listing non-goals helps to focus discussion and make progress. -->
+
+## Proposal
+
+<!-- This is where we get down to the nitty gritty of what the proposal actually is. -->
+
+Proposed approach:
+- Define an interface for the plugin to register itself with the dashboard which the plugin should implement.
+  - Something like a `PluginService` instance can expose dashboard utilities to the plugin.
+- Generate and distribute the plugin bundle.
+- Mount the generated plugin bundle on demand in a dashboard view.
+
+How can a plugin be registered?
+  - The dashboard could periodically scan `plugins` directory mounted into the pod for `YAMLs`, if there is a new plugin, register it with the dashboard.
+
+How can a plugin connect to external APIs?
+- `HttpClient` can be exposed via the `PluginService` instance for the plugin to use and make network calls.
```

Comment:
In case plugin would want to connect to external APIs I'd rather make it use some backend written in Go instead of struggling with CORS, etc.

Reply:
So the requests would go through the `kube-api-server`? Sounds good, but in that case, we'll need to expose it for the plugin to utilise.